### PR TITLE
feat: create 'GuildTemplate' & add 'Client#fetchGuildTemplate()'

### LINKED ISSUE #4651: Idea: Server templates
It would be really helpful for developers of security bots (like me) to be able to create and update server templates in a guild to create backups of servers - similar to how discordpy offers [discord.Template](https://discordpy.readthedocs.io/en/latest/api.html#template).

### PR DESCRIPTION
**Pull request by izexi**
- Ref: https://github.com/discord/discord-api-docs/pull/1712
 - Adds `DataResolver#resolveGuildTemplateCode()` to resolve template URLs
 - Along with `DataResolver#resolveCode()` to reuse code for the above and the `resolveInviteCode` method
 - Creates a `GuildTemplate` structure with a `createGuild` method to add support for the `POST /guilds/templates/:template_code` endpoint
 - Adds `Client#fetchGuildTemplate()` to support for the `GET /guilds/templates/:template_code` endpoint
 - Resolves #4651 

<details>
<summary>Here's a snippet of how I tested these changes</summary>

```js
import { once } from 'events';
import { Client, DataResolver, GuildTemplate } from 'discord.js';
import { TEST_TEMPLATE_CODE, TOKEN } from './constants';

const client = new Client();
let createdGuild;

client.login(TOKEN);

test('DataResolver#resolveGuildTemplateCode()', () => {
  expect(
    DataResolver.resolveGuildTemplateCode(
      'https://discord.com/template/FKvmczH2HyUf'
    )
  ).toBe('FKvmczH2HyUf');

  expect(
    DataResolver.resolveGuildTemplateCode('https://discord.new/FKvmczH2HyUf')
  ).toBe('FKvmczH2HyUf');

  expect(
    DataResolver.resolveGuildTemplateCode('https://discord.invalid/code')
  ).toBe('https://discord.invalid/code');
});


test('Client#fetchGuildTemplate() & GuildTempalte#createGuild()', async () => {
  await once(client, 'ready');

  const template = await client.fetchGuildTemplate(TEST_TEMPLATE_CODE);
  expect(template).toBeInstanceOf(GuildTemplate);

  const createdGuildName = 'foo';
  createdGuild = await template.createGuild(createdGuildName);
  expect(createdGuild.name).toBe(createdGuildName);
});

afterAll(() => createdGuild.delete());
```

</details>

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/util/Constants.js [hunk 1]
```diff
@@ -84,6 +84,7 @@ exports.DefaultOptions = {
     api: 'https://discord.com/api',
     cdn: 'https://cdn.discordapp.com',
     invite: 'https://discord.gg',
+    template: 'https://discord.com/template',
```
Comment (NotSugden) [inline src/util/Constants.js:87]: there needs to be a jsdoc `@property` tag for this too, and the typings need updating for it

**Review by NotSugden — CHANGES_REQUESTED** [review]

File: typings/index.d.ts [hunk 2]
```diff
@@ -903,6 +906,24 @@ declare module 'discord.js' {
     public toString(): string;
   }
 
+  export class GuildTemplate extends Base {
+    constructor(client: Client, data: object);
+    private _patch(data: object): void;
```
Comment (NotSugden) [inline typings/index.d.ts:911]: _patch isn't in the typings at all, would be inconsistent to only have it here

**Review by NotSugden — CHANGES_REQUESTED** [review]

**Review by NotSugden — CHANGES_REQUESTED** [review]
json error code 10057 unknown template could be added to this PR https://discord.com/developers/docs/topics/opcodes-and-status-codes#json

**Comment by izexi** [comment]
Closing this PR since its being superseded by #4907 and the referenced API docs PR on there has been merged

**Pull request closed without merging**